### DISCUSSION TITLE
Shutdown mechanism for recurring `ExecutionPlan`

### DISCUSSION THREAD

**Discussion by farhin23** [eclipse-edc/FederatedCatalog#273] [Ideas]
# Shutdown mechanism for recurring `ExecutionPlan`
This proposal aims to introduce mechanism to stop recurring task of `ExecutionPlan` and shutdown its executor service. 

# Why we want this?
When terminating a BaseRuntime that uses a recurring  `ExecutionPlan`, it is essential to ensure that the associated executor service is properly shut down.
While the executor service threads will eventually terminate when the java process exits, relying solely on this behavior can lead to unnecessary resource consumption. The proposed mechanism can trigger and ensure a clean shutdown of these threads.

# Solution Proposal
A `ExecutionPlan#stop()` method can be introduced. The execution plans that initiate a recurring task can override the `stop()` method to implement their own shutdown process. For execution plans with non-recurring tasks, this can be a no-op method.

# Affected Areas
* `ExecutionPlan` interface, 
* `RecurringExecutionPlan` (implements `ExecutionPlan`)

**Comment by ndr-brt** [comment]
We need to add the `stop` method both on `ExecutionPlan` (and its implementations) and on the `ExecutionManager`.

The `stop` method in the `ExecutionManager` will be called by the extension during the `shutdown` phase, and it will call the `stop` method on the `ExecutionPlan`.

Looks pretty straightforward :)

**Reply by farhin23** [reply]
> We need to add the `stop` method both on `ExecutionPlan` (and its implementations)

I agree.

> and on the `ExecutionManager`.
> The `stop` method in the `ExecutionManager` will be called by the extension during the `shutdown` phase

As far as I understand the `ExecutionManager` is not injected/registered in the EDC runtime context and therefore we can not `@Inject` it in our extension classes. 
In case of a `ExecutionManager#stop()`, we have to trigger it from the `FederatedCatalogCacheExtension` where the `ExecutionManager` gets created and used to start the `ExecutionPlan`. So in this case we have to add `shutdown` method (`@Override` shutdown of ServiceExtension) in `FederatedCatalogCacheExtension` and invoke `ExecutionManager#stop()` from there. 

Also, the `ExecutionPlan` gets injected in the EDC runtime. So the `ExecutionPlan#stop()` can be directly triggered from the extension classes.

Please let me know if I have understood properly or missed something.

**Reply by ndr-brt** [reply]
the `ExecutionPlan` gets registered only because like this it can be overridden, in fact the `RecurringExecutionPlan` is registered by default:
https://github.com/eclipse-edc/FederatedCatalog/blob/09efe6556af5ac261562513876699ac945d6655e/core/federated-catalog-core/src/main/java/org/eclipse/edc/catalog/cache/FederatedCatalogDefaultServicesExtension.java#L81-L99

But, at the end, the `ExecutionPlan` is ran by the `ExecutionManager`:
https://github.com/eclipse-edc/FederatedCatalog/blob/09efe6556af5ac261562513876699ac945d6655e/core/crawler-core/src/main/java/org/eclipse/edc/catalog/cache/ExecutionManager.java#L71-L75

So it should be stopped by the same `ExecutionManager`, with an additional `stop` method as we said.
Then the `ExecutionPlan` is started by the `FederatedCatalogCacheExtension`:
https://github.com/eclipse-edc/FederatedCatalog/blob/09efe6556af5ac261562513876699ac945d6655e/core/federated-catalog-core/src/main/java/org/eclipse/edc/catalog/cache/FederatedCatalogCacheExtension.java#L138-L141

so the shutdown hook must be set there.

**Reply by farhin23** [reply]
> So it should be stopped by the same `ExecutionManager`, with an additional `stop` method as we said

Got it. 

> The `stop` method in the `ExecutionManager` will be called by the extension during the `shutdown` phase

So we need the `FederatedCatalogCacheExtension#shutdown()` which will call `ExecutionManager#stop()` during the `shutdown` phase. Right?

**Reply by ndr-brt** [reply]
exactly. Every extension that starts an async process should be responsible to stop it down (e.g. see `TransferCoreExtension` or `ContractCoreExtension` in the connector repo)

**Reply by farhin23** [reply]
Okay. I will create an Issue and PR.
